File: LeetCode/Problems/Python/143. Reorder List.py
#Time: O(n)
#Space: O(1)
class Solution:
    def reorderList(self, head: ListNode) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if not head or not head.next:
            return
        fast = head
        slow = head
        #reverse the right part of the linked list
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
                  
        #after this, l.next will be the first node of the list to be reverse
        prev = None
        curr = slow.next
        slow.next = None

        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
            
        #we'll get prev as the head of the reversed list
        first = head
        second = prev
        while second:
            temp = first.next
            first.next = second
            second = second.next
            first.next.next = temp
            first = temp

File: LeetCode/Problems/C++/143. Reorder List.cpp
//Time: O(n)
//Space: O(1)
class Solution {
public:
    void reorderList(ListNode* head) {
        if(head == NULL) return;
        
        //step 1: find the mid point
        ListNode* fast = head->next;
        ListNode* slow = head;
        
        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
        }
        //slow will be the node before the second part
        
        //step 2: reverse the second part
        ListNode* prev = NULL;
        ListNode* curr = slow->next;
        //break the link
        slow->next = NULL;
        
        while(curr){
            ListNode* temp = curr->next;
            curr->next = prev;
            //take a step
            prev = curr;
            curr = temp;   
        }
        
        ListNode* node1 = head;
        ListNode* node2 = prev;
        //step 3: merge two list
        while(node1 && node2){
            ListNode* temp1 = node1->next; 
          
            
            node1->next = node2;
            //take a step
            node2 = node2->next;
            node1->next->next = temp1;
            node1 = temp1;

        }

    }
};